lua/cmp/utils/window.lua

```diff
@@ -204,65 +223,60 @@ window.visible = function(self)
-    width = self.style.width + border_width + (has_scrollbar and 1 or 0),
-    height = self.style.height,
-    border_width = border_width,
-    has_scrollbar = has_scrollbar,
+    width = self.style.width + border_width,
+    height = self.style.height + border_height,
+    -- NOTE: this is the scrollbar THUMB information.
+    border = {
+      height = border_height,
+      width = border_width,
+    },
   }
-end
 
----Get border width
----@return number
-window.get_border_width = function(self)
-  local border = self.style.border
-  if type(border) == 'table' then
-    local new_border = {}
-    while #new_border < 8 do
-      for _, b in ipairs(border) do
-        table.insert(new_border, b)
+  -- Information about the scrollbar
+  if self.win and type(self.scrollbar) == 'string' then
+    local content_height = self:get_content_height()
+    if content_height > self.style.height then
+      info.scrollbar = {
+        height = math.ceil(self.style.height * (self.style.height / content_height)),
+        width = 1,
+      }
+      info.scrollbar.col = info.col + info.width - ((self.style.border[4] == '' or self.style.border[4] == ' ') and 0 or 1)
+      info.scrollbar.row = info.row +
+        math.min(self.style.height - info.scrollbar.height, math.floor(self.style.height * (vim.fn.getwininfo(self.win)[1].topline / content_height))) +
+        ((self.style.border[2] == '' or self.style.border[2] == ' ' or self.style.border == 'shadow') and 0 or 1)
+      if border_width < 1 then
+        info.width = info.width + info.scrollbar.width
       end
     end
-    border = new_border
   end
 
-  local w = 0
-  if border then
-    if type(border) == 'string' then
-      if border == 'single' then
-        w = 2
-      elseif border == 'solid' then
-        w = 2
-      elseif border == 'double' then
-        w = 2
-      elseif border == 'rounded' then
-        w = 2
-      elseif border == 'shadow' then
-        w = 1
-      end
-    elseif type(border) == 'table' then
-      local b4 = type(border[4]) == 'table' and border[4][1] or border[4]
-      if #b4 > 0 then
-        w = w + 1
-      end
-      local b8 = type(border[8]) == 'table' and border[8][1] or border[8]
-      if #b8 > 0 then
-        w = w + 1
-      end
-    end
+  return info
+end
+
+--- @return number height, number width
+--- @return number[] dimensions the height and width
+window.get_border_dimensions = function(self)
+  local border = self.style.border
+
+  if border == 'shadow' then
+    return 1, 1
+  elseif type(border) == 'string' then
+    return 2, 2
+  elseif type(border) == 'table' then
+    -- NOTE: the border indices look like this: 1 2 3
+    --                                          8   4
+    --                                          7 6 5
+    return (self.style.border[2] ~= '' and 1 or 0) + (self.style.border[6] ~= '' and 1 or 0),
+           (self.style.border[4] ~= '' and 1 or 0) + (self.style.border[8] ~= '' and 1 or 0)
   end
```

Comment:
I think we should fill to 8 characters like the following way.

```
        local normalized_border = {}
        while #normalized_border < 8 do
          for _, b in ipairs(self.style.border) do
            table.insert(normalized_border, b)
          end
        end
```

Reply:
I did some testing before this implementation. As I understand it, this is how the Nvim API behaves:

* `{'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'}` = full border
* `{'', 'b', '', 'd', '', 'f', '', 'h'}` = helpful error
* `{'a', '', 'c', '', 'e', '', 'g', ''}` = no border

So only indices 2, 4, 6, and 8 actually indicate height and width. I think the error provided by Neovim is helpful as well: 

> corner between used edges must be specified

It's one less thing we have to compute by not worrying about it

Reply:
Seems like 2, 4, and 8 items are all possible. I'll account for that